SpiceSharp/Simulations/Base/BaseSimulation.cs

```diff
@@ -0,0 +1,591 @@
+
+        /// <summary>
+        /// Calculate the operating point of the circuit
+        /// </summary>
+        /// <param name="maxIterations">Maximum iterations</param>
+        protected void Op(int maxIterations)
+        {
+            var state = RealState;
+            var config = BaseConfiguration;
+            state.Init = RealState.InitializationStates.InitJunction;
+            state.Matrix.Complex = false;
+
+            // First, let's try finding an operating point by using normal iterations
+            if (!config.NoOperatingPointIterations)
+            {
+                if (Iterate(maxIterations))
+                    return;
+            }
+
+            // No convergence, try Gmin stepping
+            if (config.GminSteps > 1)
+            {
+                state.Init = RealState.InitializationStates.InitJunction;
+                CircuitWarning.Warning(this, "Starting Gmin stepping");
+                state.DiagonalGmin = config.Gmin;
+                for (int i = 0; i < config.GminSteps; i++)
+                    state.DiagonalGmin *= 10.0;
+                for (int i = 0; i <= config.GminSteps; i++)
+                {
+                    state.IsConvergent = false;
+                    if (!Iterate(maxIterations))
+                    {
+                        state.DiagonalGmin = 0.0;
+                        CircuitWarning.Warning(this, "Gmin step failed");
+                        break;
+                    }
+                    state.DiagonalGmin /= 10.0;
+                    state.Init = RealState.InitializationStates.InitFloat;
+                }
+                state.DiagonalGmin = 0.0;
+                if (Iterate(maxIterations))
+                    return;
+            }
+
+            // Nope, still not converging, let's try source stepping
+            if (config.SourceSteps > 1)
+            {
+                state.Init = RealState.InitializationStates.InitJunction;
+                CircuitWarning.Warning(this, "Starting source stepping");
+                for (int i = 0; i <= config.SourceSteps; i++)
+                {
+                    state.SourceFactor = i / (double)config.SourceSteps;
+                    if (!Iterate(maxIterations))
+                    {
+                        state.SourceFactor = 1.0;
+                        // circuit.CurrentAnalysis = AnalysisType.DoingTran;
+                        CircuitWarning.Warning(this, "Source stepping failed");
+                        return;
+                    }
+                }
+                state.SourceFactor = 1.0;
+                return;
+            }
+
+            // Failed
+            throw new CircuitException("Could not determine operating point");
+        }
+
+        /// <summary>
+        /// Solve iteratively for simulations
+        /// </summary>
+        /// <param name="maxIterations">Maximum number of iterations</param>
+        /// <returns></returns>
+        protected bool Iterate(int maxIterations)
```

Comment:
@svenboulanger Consider refactor this long method

Reply:
We may review this at a later date. The method is a copy from Spice 3f5 and will probably be subject to change later anyway.

Reply:
I have turned it into issue #24